### ISSUE TITLE
Adaption of web2.iweb interfaces is broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2199]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#2199|
|Type|defect|
|Created|2006-10-27 18:41:15Z|

As of r18455, adaption of web2.iweb interfactes is broken in that it no longer raises ```TypeError``` when adaptions fails and instead returns ```None```.

A change to ```SpecialAdaptInterfaceClass.__call__()``` appears to be the cause.

<details><summary>Searchable metadata</summary>

```
trac-id__2199 2199
type__defect defect
reporter__wsanchez wsanchez
priority__highest highest
milestone__Web2_Gold_Master Web2-Gold-Master
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1161974475000000 1161974475000000
changetime__1163436563000000 1163436563000000
version__None None
owner__ 
cc__wsanchez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_1"></a>|@wsanchez commented|
|-|-|

Perhaps something more like:
```
# Is there a better way to do this than this funky extra class?
NotSpecified = object()
class SpecialAdaptInterfaceClass(interface.InterfaceClass):
    # A special adapter for IResource to handle the extra step of adapting
    # from IOldNevowResource-providing resources.
    def __call__(self, other, alternate=NotSpecified):
        if alternate == NotSpecified:
            try:
                return super(SpecialAdaptInterfaceClass, self).__call__(other)
            except TypeError:
                pass
        else:
            result = super(SpecialAdaptInterfaceClass, self).__call__(other, alternate)
            if result is not alternate:
                return result
        
        if alternate == NotSpecified:
            return IOldNevowResource(other)
        result = IOldNevowResource(other, alternate)
        if result is not alternate:
            return IResource(result)
        return alternate
IResource.__class__ = SpecialAdaptInterfaceClass
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Ready for review in transitive-adaption-2199

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

Looks good, greatly improves test coverage. I was only wondering if the tests wouldn't be better in test_compat than in test_server. Anyway it should be merged.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [18666]) Merge transitive-adaption-2199

Author: exarkun
Reviewer: therve
Fixes [#2199](https://github.com/twisted/twisted/issues/2199)

Add tests for various adaption behaviors related to IResource and fix
the implementation of IResource.__class__.__call__ to make these new
tests pass.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
